src/WebJobs.Script.Analyzers/AssemblyCache.cs

```diff
@@ -0,0 +1,240 @@
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Azure.WebJobs;
+using Microsoft.Azure.WebJobs.Host;
+using Microsoft.Azure.WebJobs.Hosting;
+using Microsoft.CodeAnalysis;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+
+namespace Microsoft.Azure.Functions.Analyzers
+{
+    // Only support extensions and WebJobs core.
+    // Although extensions may refer to other dlls.
+    public class AssemblyCache
+    {
+        // Map from assembly identities to full paths
+        public static AssemblyCache Instance = new AssemblyCache();
+
+        bool _registered;
+
+        // Assembly Display Name --> Path
+        Dictionary<string, string> _assemblyNameToPathMap = new Dictionary<string, string>();
+
+        // Assembly Display Name --> loaded Assembly object
+        Dictionary<string, Assembly> _assemblyNameToObjMap = new Dictionary<string, Assembly>();
+
+        const string WebJobsAssemblyName = "Microsoft.Azure.WebJobs";
+        const string WebJobsHostAssemblyName = "Microsoft.Azure.WebJobs.Host";
+
+        JobHostMetadataProvider _tooling;
+
+        internal JobHostMetadataProvider Tooling => _tooling;
+        private int _projectCount;
+
+        // $$$ This can get invoked multiple times concurrently
+        // This will get called on every compilation.
+        // So return early on subsequent initializations.
+        internal void Build(Compilation compilation)
+        {
+            Register();
+
+            int count;
+            lock (this)
+            {
+                // If project references have changed, then reanalyze to pick up new dependencies.
+                var refs = compilation.References.OfType<PortableExecutableReference>().ToArray();
+                count = refs.Length;
+                if ((count == _projectCount) && (_tooling != null))
+                {
+                    return; // already initialized.
+                }
+
+                // Even for netStandard/.core projects, this will still be a flattened list of the full transitive closure of dependencies.
+                foreach (var asm in compilation.References.OfType<PortableExecutableReference>())
+                {
+                    var dispName = asm.Display; // For .net core, the displayname can be the full path
+                    var path = asm.FilePath;
+
+                    _assemblyNameToPathMap[dispName] = path;
+                }
+
+                // Builtins
+                _assemblyNameToObjMap["mscorlib"] = typeof(object).Assembly;
+                _assemblyNameToObjMap[WebJobsAssemblyName] = typeof(Microsoft.Azure.WebJobs.FunctionNameAttribute).Assembly;
+                _assemblyNameToObjMap[WebJobsHostAssemblyName] = typeof(Microsoft.Azure.WebJobs.JobHost).Assembly;
+
+                // JSON.Net?
```

Comment:
Is this a todo?

Reply:
This was part of Mike's original code. It looks like a todo, but for what, I'm not sure without a trip into the institutional memory stacks. It may be outdated since this code was written pre-V2 Functions, but in the interest of not being That Translator Who Went Overboard, I left it in as a breadcrumb in case it's useful if we make future enhancements to these analyzers.